kong/kong.lua

```diff
@@ -157,6 +161,34 @@ function Kong.ssl_certificate()
   end
 end
 
+function Kong.balancer()
+  local addr = ngx.ctx.balancer_address
+  addr.tries = addr.tries + 1
+  if addr.tries > 1 then  
+    -- only call balancer on retry, first one is done in `core.access.before` which runs
+    -- in the ACCESS context and hence has less limitations than this BALANCER context
+    -- where the retries are executed
+    
+    -- record failure data
+    addr.failures = addr.failures or {}
+    local state, code = ngx_balancer.get_last_failure()
+    addr.failures[addr.tries-1] = { name = state, code = code }
```

Comment:
What to do with those? I believe the question before that would be: how many tries do we allow? It seems the default `proxy_next_upstream_tries` value is 0, which doesn't set any limit to the number of tries? Will we go on forever if none of the upstreams are responsive?

If/when we hit our - configurable, ideally - limit, then we should probably log those failures.